website/static/js/statistics.js

```diff
@@ -2,241 +2,378 @@
 
 require('keen-dataviz/dist/keen-dataviz.min.css');
 
+var oop = require('js/oop');
 var $osf = require('js/osfHelpers');
 var keenDataviz = require('keen-dataviz');
 var keenAnalysis = require('keen-analysis');
 
-var ProjectUsageStatistics = function(keenProjectId, keenReadKey){
-    var self = this;
+
+var UserFacingChart = oop.defclass({
+    constructor: function(params) {
+        var self = this;
+        params = params || {};
+
+        var required = ['keenProjectId', 'keenReadKey', 'containingElement'];
+        required.forEach(function(paramName) {
+            if (!params[paramName]) {
+                throw 'Missing required argument "' + paramName + '"';
+            }
+        });
+
+        /**
+         * Manages the connections to Keen.  Needs a project id and valid read key for that project.
+         *
+         * @property keenClient
+         * @type {keenAnalysis}
+         */
+        self.keenClient = new keenAnalysis({
+            projectId: params.keenProjectId,
+            readKey: params.keenReadKey,
+        });
+
+        /**
+         * HTML id of containing element.
+         *
+         * @property containingElement
+         * @type {string}
+         */
+        self.containingElement = params.containingElement;
+
+        /**
+         * Max number of entries to display for category-based charts. Ignored for time-series data.
+         * Default 10.
+         *
+         * @property MAX_DISPLAY_ENTRIES
+         * @type {integer}
+         */
+        self.MAX_DISPLAY_ENTRIES = params.maxDisplayEntries || 10;
+    },
+
+    /**
+     * Returns an Object that defines the analysis to relay to Keen. Has two top-level keys,
+     * `type` and `params`. `type` is the type of aggregation to perform. `params` defines the
+     * parameters of the query.  The timeframe for the analysis is be filled in by the query()
+     * method, which calls baseQuery().
+     *
+     * See: https://keen.io/docs/api/?javascript#analyses
+     *
+     * @method baseQuery
+     * @return {Object}
+     */
+    baseQuery: function() {
+        throw 'Concrete class does not provide a "baseQuery" method!';
```

Comment:
Minor: As per COSDev, always throw an instance an instance of `Error`. This ensures that the error will get picked up by Raven.

```js
throw new Error('Concrete class..')
```